### PR TITLE
Initial introduction and implementation of `oidc.IDTokenSource`

### PR DESCRIPTION
**Pull request by dekkagaijin**
This PR introduces the `oidc.IDTokenSource` concept and refactors the existing `oauthflow.InteractiveIDTokenGetter` as part of the overall SDK work: https://github.com/sigstore/cosign/issues/844

We want to conceptually isolate the task of retrieving Open ID Tokens, I've introduced the concept of an `IDTokenSource`. This is a prerequisite for implementing a sane Fulcio client for `cosign`, and unblocking higher level work.

```release-note
Introduced `oidc.IDTokenSource` to abstract the generation of Open ID tokens
```

### CODE REVIEW

File: pkg/oauth/oidc/interactive.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+	cfg.RedirectURL = oobRedirectURI
+	authURL := cfg.AuthCodeURL(stateToken, opts...)
+	fmt.Fprintln(os.Stderr, "Go to the following link in a browser:\n\n\t", authURL)
+	fmt.Fprintf(os.Stderr, "Enter verification code: ")
+	var code string
+	fmt.Scanln(&code)
+	return code
+}
+
+func startRedirectListener(state, htmlPage string, doneCh chan string, errCh chan error) (*http.Server, *url.URL, error) {
```
Comment (mattmoor) [inline pkg/oauth/oidc/interactive.go:48]: Generally when I see `doneCh` I think `chan struct{}{}`, maybe `codeCh` is a better name?

File: pkg/oauth/oidc/interactive.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+	listener, err := net.Listen("tcp", "localhost:0")
```
Comment (mattmoor) [inline pkg/oauth/oidc/interactive.go:49]: `:0` is probably worth a comment.  I'm guessing it means "pick a port"?

Comment (mattmoor) [inline pkg/oauth/oidc/interactive.go:49]: Hmm, this is never closed 🤔

File: pkg/oauth/oidc/interactive.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+		if r.FormValue("state") != state {
+			errCh <- errors.New("invalid state token")
+			return
+		}
+		doneCh <- r.FormValue("code")
+		fmt.Fprint(w, htmlPage)
+	})
+
+	go func() {
+		if err := s.Serve(listener); err != nil && err != http.ErrServerClosed {
```
Comment (mattmoor) [inline pkg/oauth/oidc/interactive.go:80]: errors.Is?

File: pkg/oauth/oidc/interactive.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+func getCode(doneCh chan string, errCh chan error) (string, error) {
+	timeoutCh := time.NewTimer(120 * time.Second)
+	select {
+	case code := <-doneCh:
+		return code, nil
+	case err := <-errCh:
+		return "", err
+	case <-timeoutCh.C:
```
Comment (mattmoor) [inline pkg/oauth/oidc/interactive.go:95]:
```suggestion
	case <-time.After(120 * time.Second):
```

File: pkg/oauth/oidc/interactive.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+	return ksuid.New().String()
+}
+
+func doInteractiveIDTokenFlow(ctx context.Context, cfg oauth2.Config, p *coreoidc.Provider, extraAuthCodeOpts []oauth2.AuthCodeOption, openBrowser browserOpener) (*IDToken, error) {
+	// generate random fields and save them for comparison after OAuth2 dance
+	stateToken := newKSUID()
+	nonce := newKSUID()
+
+	doneCh := make(chan string)
+	errCh := make(chan error)
```
Comment (mattmoor) [inline pkg/oauth/oidc/interactive.go:111]: Are these ever closed?

**Review by mattmoor — COMMENTED** [review]
Generally LGTM, but noticed a bunch of random stuff.

We should probably get golangci-lint running here with all the checks we have in cosign

File: pkg/oauth/oidc/interactive.go (hunk 2 above)
Comment (dekkagaijin) [reply pkg/oauth/oidc/interactive.go]: It's closed in this perfectly clear block of `doInteractiveIDTokenFlow` :p

```
defer func() {
  go func() {
    _ = redirectServer.Shutdown(context.Background())
  }()
}()
```

File: pkg/oauth/oidc/interactive.go (hunk 3 above)
Comment (dekkagaijin) [reply pkg/oauth/oidc/interactive.go]: done

File: pkg/oauth/oidc/interactive.go (hunk 4 above)
Comment (dekkagaijin) [reply pkg/oauth/oidc/interactive.go]: done

File: pkg/oauth/oidc/interactive.go (hunk 5 above)
Comment (dekkagaijin) [reply pkg/oauth/oidc/interactive.go]: they are now!

**Comment by dekkagaijin** [comment]
> We should probably get golangci-lint running here with all the checks we have in cosign

I'll take a look and see how they differ

**Comment by dekkagaijin** [comment]
PR to update golangci-lint: https://github.com/sigstore/sigstore/pull/322

**Review by mattmoor — APPROVED** [review]

File: pkg/oauth/oidc/interactive.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+type interactiveIDTokenSource struct {
+	cfg               oauth2.Config
+	oidp              *coreoidc.Provider
+	extraAuthCodeOpts []oauth2.AuthCodeOption
+	browser           browserOpener
+}
+
+var errWontOpenBrowser = errors.New("not opening that browser")
```
Comment (bobcallaway) [inline pkg/oauth/oidc/interactive.go:168]: can we make this message a bit more user-friendly? Is there some way that the user can change this (e.g. verify the default system-wide browser is operational?)

**Review by bobcallaway — COMMENTED** [review]
i only have one nit on an error message, but the bigger question i have is whether the existing e2e tests need to be updated to see that this new package works correctly?

File: pkg/oauth/oidc/interactive.go (hunk 6 above)
Comment (dekkagaijin) [reply pkg/oauth/oidc/interactive.go:168]: This won't ever be surfaced to a user- it only exists to force this flow to use command line inputs over browser redirects

**Comment by dekkagaijin** [comment]
> but the bigger question i have is whether the existing e2e tests need to be updated to see that this new package works correctly?

The existing e2e tests don't exercise any of the code that I moved around to here. I've written tests for this flow before, but I'll need to do a more intrusive refactor in order to make the current code testable.

**Comment by dekkagaijin** [comment]
never mind, found, copied, and adapted the e2e test

File: pkg/oauth/oidc/interactive.go (hunk 6 above)
Comment (bobcallaway) [reply pkg/oauth/oidc/interactive.go:168]: if someone called `InteractiveIDTokenSource` with `allowBrowser=false`, it seems like line 142 would print this message; I could be misreading it though.

File: pkg/oauth/oidc/interactive.go [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+		return "", errors.New("timeout")
+	}
+}
+
+// newKSUID returns a globally unique, base62 (URL-safe) encoded, 27 character string.
+func newKSUID() string {
+	return ksuid.New().String()
+}
+
+func doInteractiveIDTokenFlow(ctx context.Context, cfg oauth2.Config, p *coreoidc.Provider, extraAuthCodeOpts []oauth2.AuthCodeOption, openBrowser browserOpener) (*IDToken, error) {
```
Comment (bobcallaway) [inline pkg/oauth/oidc/interactive.go:104]: is there a reason not to make this a method on `interactiveIDTokenSource`? all the args except for `ctx` end up being private member variables of the same struct?

Comment (dekkagaijin) [reply pkg/oauth/oidc/interactive.go]: done

**Review by bobcallaway — APPROVED** [review]

**Pull request merged**